go/provisioner/local/pgbackrest/README.md

```diff
@@ -6,97 +6,243 @@ pgBackRest has specific ways of doing things and handles many backup-related tas
+   - A shared server certificate and key (`<backup_repo>/tls/server.crt`,
+     `server.key`)
+
+2. Each `pgbackrest.conf` includes:
+   - TLS server configuration (`tls-server-cert-file`, `tls-server-key-file`,
+     `tls-server-ca-file`, `tls-server-address`, `tls-server-port`)
+   - TLS client configuration for connecting to other nodes
+     (`pgN-host-type=tls`, `pgN-host-port`, `pgN-host-ca-file`,
+     `pgN-host-cert-file`, `pgN-host-key-file`)
+   - Server authorization (`tls-server-auth=pgbackrest=*`) allowing clients with certificate CN "pgbackrest" to execute any command
+
+3. During database provisioning, a `pgbackrest-server` service is started for each cell via `StartPgBackRestServer()`.
+
+4. When backing up from a standby, the pgBackRest client connects to the primary's TLS server to fetch required files.
+
+Consult the [pgBackRest server reference](https://pgbackrest.org/configuration.html#section-server) for full details on configuring the TLS server.
+
+### Port allocation
+
+pgBackRest TLS servers use port 8432 + cell_index (cells sorted alphabetically, 0-indexed):
+
+- zone1 (index 0): 8432
+- zone2 (index 1): 8433
+- zone3 (index 2): 8434
+
+The base port is defined in `ports.DefaultPgBackRestTLSPort`.
+
+### Certificate generation
+
+The `tls.go` file provides two functions for certificate generation:
+
+- `GenerateCA()` - Creates a self-signed ECDSA P256 CA certificate (valid for
+  10 years)
+- `GenerateServerCert()` - Creates a server certificate signed by the CA with
+  ServerAuth and ClientAuth extended key usage (valid for 1 year)
+
+The server certificate uses CN "pgbackrest" and includes SANs for `localhost`, `127.0.0.1`, and `::1`.
+
+## Summary of pgbackrest.conf generation rules
+
+### Every pooler has its own `pgbackrest.conf`
+
+- Its own Postgres instance is listed first (as `pg1-*`), because that's the
+  implicit target of restores and the first choice for backup source.
+- The other poolers have entries (`pg2-*`, `pg3-*`, etc.) are needed because of
+  the potential for inter-pooler TCP connections.
+
+### Backing up a primary database
+
+When backing up a primary (e.g. during bootstrap), pgBackRest does not need to
+connect to other poolers. However, it will _attempt_ to connect to other
+poolers to verify replication status, etc. So, if during bootstrap, if the
+standbys are potentially not available, the backup may fail or timeout. A fix
+for this (implemented in some test code) is to initially create a
+`pgbackrest.conf` that does not reference the standbys. Then, after bootstrap
+has completed, the primary would then have a full `pgbackrest.conf` file.
+
+### Backing up a standby database
+
+A pgBackRest process on the standby initiates two connections to the primary pooler node:
+
+1. pgBackRest TLS server: used for retrieving `pg_control` and other files that
+   must come from the primary
+2. Postgres server port: for issuing `pg_start_backup()`
```

Comment:
it's now `pg_backup_start()` (renamed in PG15)

Reply:
Indeed. This is a surprise to me. Reminds years ago, when non-exclusive backup wasn't possible -- then we did need to deal with primary.

this adds dependency I don't like. We basically need two hosts to be alive during backup.

would be good to understand what's this, old, non-updated decision in pgBackRest (would be another surprise) or there is a big reason for it

wal-g does full backups from standbys not involving primary

Reply:
Found some answers here:

https://github.com/pgbackrest/pgbackrest/issues/2454#issuecomment-2377204575

https://pgstef.github.io/2024/09/23/pgbackrest_backups_from_the_standby_server.html